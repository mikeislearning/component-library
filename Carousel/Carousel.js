// @format

import React from 'react';
import { Image, Text, View } from 'react-native';
import { PropTypes } from 'prop-types';
import Carousel, { Pagination } from 'react-native-snap-carousel';

import { Button, Spacer } from 'app/components';
import { colors, metrics } from 'app/theme';

import styles from './Carousel.styles';

export default class CarouselComponent extends React.Component {
  static propTypes = {
    entries: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string,
        image: PropTypes.number,
        subtitle: PropTypes.subtitle,
        button: PropTypes.shape({
          text: PropTypes.string,
          theme: PropTypes.string,
          onPress: PropTypes.func,
        }),
        smallButton: PropTypes.shape({
          text: PropTypes.string,
          onPress: PropTypes.func,
        }),
      })
    ),
    fullWidth: PropTypes.bool,
  };

  state = { dotIndex: 0 };

  renderItem = ({ item }) => {
    const { fullWidth } = this.props;
    const imageStyles = fullWidth ? styles.fullImage : styles.image;
    const widthStyle = fullWidth ? styles.fullText : styles.modalText;

    const titleStyles = [widthStyle, styles.carouselText, styles.title];
    const subtitleStyles = [widthStyle, styles.carouselText, styles.subtitle];
    const textStyles = [widthStyle, styles.carouselText, styles.text];

    const resizeMode = fullWidth ? 'cover' : 'contain';

    return (
      <View style={styles.item}>
        <Text style={titleStyles}>{item.title}</Text>
        <Image source={item.image} style={imageStyles} resizeMode={resizeMode} />
        {item.subtitle && <Text style={subtitleStyles}>{item.subtitle}</Text>}
        {item.text && <Text style={textStyles}>{item.text}</Text>}
        {item?.button && (
          <React.Fragment>
            <Spacer size="small" />
            <Button fullWidth={true} onPress={item.button?.onPress} theme={item.button?.theme}>
              {item?.button?.text}
            </Button>
          </React.Fragment>
        )}
        {item?.smallButton && (
          <React.Fragment>
            <Spacer size="small" />
            <Button onPress={item.smallButton?.onPress} theme="transparent">
              {item.smallButton?.text}
            </Button>
          </React.Fragment>
        )}
      </View>
    );
  };

  render() {
    const { entries, fullWidth } = this.props;

    const width = fullWidth ? metrics.screenWidth : 300;

    return (
      <View style={styles.container}>
        <Carousel
          data={entries}
          itemWidth={width}
          onSnapToItem={(index) => this.setState({ dotIndex: index })}
          renderItem={this.renderItem}
          sliderWidth={width}
        />
        <Pagination
          dotsLength={entries.length}
          activeDotIndex={this.state.dotIndex}
          dotColor={colors.arriveBlack}
          inactiveDotColor={colors.gray.medium}
        />
      </View>
    );
  }
}
