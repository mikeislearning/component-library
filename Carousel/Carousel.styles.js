import { StyleSheet } from 'react-native';

import { colors, fonts, metrics } from 'app/theme';

const styles = StyleSheet.create({
  carouselText: {
    color: colors.arriveBlack,
    textAlign: 'center',
  },
  container: {
    flex: 1,
  },
  fullImage: {
    height: 250,
    marginBottom: 29,
    marginTop: 24,
    width: metrics.screenWidth,
  },
  fullText: {
    paddingHorizontal: 15,
  },
  image: {
    height: 150,
    marginBottom: 29,
    marginTop: 24,
  },
  item: {
    alignItems: 'center',
    flex: 1,
  },
  modalText: {
    paddingHorizontal: 20,
  },
  subtitle: {
    fontFamily: fonts.family.messinaSans.light,
    fontSize: fonts.size.h5,
  },
  text: {
    fontFamily: fonts.family.messinaSans.regular,
    fontSize: 15,
    lineHeight: 22,
  },
  title: {
    fontFamily: fonts.family.messinaSans.regular,
    fontSize: fonts.size.large,
  },
});

export default styles;
