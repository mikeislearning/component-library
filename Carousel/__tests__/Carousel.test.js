import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Carousel from '../Carousel';

describe('<Carousel />', () => {
  it('renders correctly', () => {
    const entries = [
      {
        title: 'title1',
        image: 32,
        subtitle: 'subtitle1',
      },
      {
        title: 'title2',
        image: 33,
        subtitle: 'subtitle2',
      },
    ];
    const renderer = new ShallowRenderer();
    renderer.render(<Carousel entries={entries} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
