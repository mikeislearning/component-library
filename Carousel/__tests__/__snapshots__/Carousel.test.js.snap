// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Carousel
    activeAnimationOptions={null}
    activeAnimationType="timing"
    activeSlideAlignment="center"
    activeSlideOffset={20}
    apparitionDelay={0}
    autoplay={false}
    autoplayDelay={1000}
    autoplayInterval={3000}
    callbackOffsetMargin={5}
    containerCustomStyle={Object {}}
    contentContainerCustomStyle={Object {}}
    data={
      Array [
        Object {
          "image": 32,
          "subtitle": "subtitle1",
          "title": "title1",
        },
        Object {
          "image": 33,
          "subtitle": "subtitle2",
          "title": "title2",
        },
      ]
    }
    enableMomentum={false}
    enableSnap={true}
    firstItem={0}
    hasParallaxImages={false}
    inactiveSlideOpacity={0.7}
    inactiveSlideScale={0.9}
    inactiveSlideShift={0}
    itemWidth={300}
    layout="default"
    lockScrollTimeoutDuration={1000}
    lockScrollWhileSnapping={false}
    loop={false}
    loopClonesPerSide={3}
    onSnapToItem={[Function]}
    renderItem={[Function]}
    scrollEnabled={true}
    shouldOptimizeUpdates={true}
    slideStyle={Object {}}
    sliderWidth={300}
    swipeThreshold={20}
    useScrollView={false}
    vertical={false}
  />
  <Pagination
    activeDotIndex={0}
    dotColor="#272E39"
    dotsLength={2}
    inactiveDotColor="#726E67"
    inactiveDotOpacity={0.5}
    inactiveDotScale={0.5}
    tappableDots={false}
    vertical={false}
  />
</View>
`;
